titus-common/src/main/java/com/netflix/titus/common/util/limiter/tokenbucket/internal/FixedIntervalTokenBucketSupplier.java

```diff
@@ -0,0 +1,147 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.limiter.tokenbucket.internal;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import com.google.common.base.Stopwatch;
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.Evaluators;
+import com.netflix.titus.common.util.ExceptionExt;
+import com.netflix.titus.common.util.limiter.tokenbucket.FixedIntervalTokenBucketConfiguration;
+import com.netflix.titus.common.util.limiter.tokenbucket.RefillStrategy;
+import com.netflix.titus.common.util.limiter.tokenbucket.TokenBucket;
+
+/**
+ * {@link TokenBucket} supplier which recreates a token bucket if any of its configurable parameters changes.
+ * The configuration parameters are read from {@link FixedIntervalTokenBucketConfiguration}.
+ */
+public class FixedIntervalTokenBucketSupplier implements Supplier<TokenBucket> {
+
+    private final String name;
+    private final FixedIntervalTokenBucketConfiguration configuration;
+    private final Consumer<TokenBucket> onChangeListener;
+    private final Optional<TitusRuntime> titusRuntime;
+
+    private volatile ActiveConfiguration activeConfiguration;
+
+    private final Object lock = new Object();
+
+    public FixedIntervalTokenBucketSupplier(String name,
+                                            FixedIntervalTokenBucketConfiguration configuration,
+                                            Consumer<TokenBucket> onChangeListener,
+                                            Optional<TitusRuntime> titusRuntime) {
+        this.name = name;
+        this.configuration = configuration;
+        this.onChangeListener = onChangeListener;
+        this.titusRuntime = titusRuntime;
+        this.activeConfiguration = reload();
+    }
+
+    @Override
+    public TokenBucket get() {
+        return isSame() ? activeConfiguration.getTokenBucket() : reload().getTokenBucket();
+    }
+
+    private boolean isSame() {
+        return activeConfiguration.getCapacity() == configuration.getCapacity()
+                && activeConfiguration.getInitialNumberOfTokens() == configuration.getInitialNumberOfTokens()
+                && activeConfiguration.getIntervalMs() == configuration.getIntervalMs()
+                && activeConfiguration.getNumberOfTokensPerInterval() == configuration.getNumberOfTokensPerInterval();
+    }
+
+    private ActiveConfiguration reload() {
+        boolean same;
+        synchronized (lock) {
+            same = activeConfiguration != null && isSame();
```

Comment:
Should the null check be also part of the isSame method?

Reply:
Good idea. I will change it as suggested.